src/common/typescript.ts

```diff
@@ -0,0 +1,23 @@
+/********************************************************************************
+ * Copyright (C) 2024 EclipseSource and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+export type TypeOfArrayElement<T> = T extends (infer E)[] ? E : T extends readonly (infer E)[] ? E : never;
```

Comment:
I believe that mutable arrays extend immutable arrays (according to TypeScript), so only the second check is required:

```suggestion
export type TypeOfArrayElement<T> = T extends readonly (infer E)[] ? E : never;
```